File: C++/023_Merge k Sorted Lists.cpp
class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        ListNode *head = NULL, *pre, *now;
        while (!pq.empty())
            pq.pop();
        for (int i = 0; i < lists.size(); i++) {
            if (lists[i] != NULL)
                pq.push(lists[i]);
        }
        while (!pq.empty()) {
            now = pq.top();
            pq.pop();
            if (head == NULL)
                head = now;
            else
                pre->next = now;
            if (now->next != NULL)
                pq.push(now->next);
            pre = now;
        }
        return head;
    }
private:
    struct cmp {
        bool operator() (const ListNode *a, const ListNode *b) {
            return a->val > b->val;
        }
    };
    priority_queue<ListNode*, vector<ListNode*>, cmp> pq;
};

File: Python/023_Merge k Sorted Lists.py
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        pq = Queue.PriorityQueue(len(lists))
        for i in lists:
            if i is not None:
                pq.put_nowait((i.val, i))
        head, pre = None, None
        while not pq.empty():
            ft = pq.get_nowait()
            if head is None:
                head = ft[1]
            else:
                pre.next = ft[1]
            pre = ft[1]
            if ft[1].next is not None:
                pq.put_nowait((ft[1].next.val, ft[1].next))
        return head